File: artifacts/yolo-service/routes/cameras.py
from datetime import datetime, timezone
from typing import List, Optional
from fastapi import APIRouter, HTTPException, Depends
from fastapi.responses import JSONResponse
from pydantic import BaseModel
from db import get_cursor
from auth_utils import get_current_user
from simulation import generate_detection_snapshot

router = APIRouter()


class CreateCameraRequest(BaseModel):
    name: str
    location: str = ""
    streamUrl: Optional[str] = None
    ppeRequirements: List[str] = []
    siteId: Optional[int] = None


def _require_editor(user: dict = Depends(get_current_user)) -> dict:
    if user["role"] not in ("admin", "support"):
        raise HTTPException(status_code=403, detail="Admin or Support role required")
    return user


def _camera_dict(row):
    d = dict(row)
    if d.get("ppe_requirements") is None:
        d["ppe_requirements"] = []
    d["ppeRequirements"] = d.pop("ppe_requirements", [])
    d["streamUrl"] = d.pop("stream_url", None)
    d["siteId"] = d.pop("site_id", None)
    d["createdAt"] = d.pop("created_at", None)
    d["updatedAt"] = d.pop("updated_at", None)
    if d["createdAt"] and hasattr(d["createdAt"], "isoformat"):
        d["createdAt"] = d["createdAt"].isoformat()
    if d["updatedAt"] and hasattr(d["updatedAt"], "isoformat"):
        d["updatedAt"] = d["updatedAt"].isoformat()
    return d


@router.get("/api/cameras")
def list_cameras():
    with get_cursor() as cur:
        cur.execute("SELECT * FROM cameras ORDER BY id")
        rows = cur.fetchall()
    return [_camera_dict(r) for r in rows]


@router.post("/api/cameras", status_code=201)
def create_camera(body: CreateCameraRequest, _user: dict = Depends(_require_editor)):
    with get_cursor(commit=True) as cur:
        cur.execute(
            """INSERT INTO cameras (name, location, stream_url, ppe_requirements, site_id)
               VALUES (%s, %s, %s, %s, %s) RETURNING *""",
            (body.name, body.location, body.streamUrl, body.ppeRequirements, body.siteId),
        )
        camera = cur.fetchone()
    return _camera_dict(camera)


@router.get("/api/cameras/{camera_id}")
def get_camera(camera_id: int):
    with get_cursor() as cur:
        cur.execute("SELECT * FROM cameras WHERE id = %s", (camera_id,))
        camera = cur.fetchone()
    if not camera:
        raise HTTPException(status_code=404, detail="Not found")
    return _camera_dict(camera)


@router.patch("/api/cameras/{camera_id}")
def update_camera(camera_id: int, body: dict, _user: dict = Depends(_require_editor)):
    allowed = {"name", "location", "status", "streamUrl", "ppeRequirements"}
    col_map = {
        "name": "name",
        "location": "location",
        "status": "status",
        "streamUrl": "stream_url",
        "ppeRequirements": "ppe_requirements",
    }

    sets = []
    vals = []
    for key, col in col_map.items():
        if key in body:
            sets.append(f"{col} = %s")
            vals.append(body[key])

    sets.append("updated_at = %s")
    vals.append(datetime.now(timezone.utc))
    vals.append(camera_id)

    with get_cursor(commit=True) as cur:
        cur.execute(
            f"UPDATE cameras SET {', '.join(sets)} WHERE id = %s RETURNING *",
            vals,
        )
        camera = cur.fetchone()
    if not camera:
        raise HTTPException(status_code=404, detail="Not found")
    return _camera_dict(camera)


@router.delete("/api/cameras/{camera_id}", status_code=204)
def delete_camera(camera_id: int, _user: dict = Depends(_require_editor)):
    with get_cursor(commit=True) as cur:
        cur.execute("DELETE FROM cameras WHERE id = %s", (camera_id,))
    return None


@router.get("/api/cameras/{camera_id}/snapshot")
def get_snapshot(camera_id: int):
    with get_cursor() as cur:
        cur.execute("SELECT * FROM cameras WHERE id = %s", (camera_id,))
        camera = cur.fetchone()
    if not camera:
        raise HTTPException(status_code=404, detail="Not found")
    snapshot = generate_detection_snapshot(dict(camera))
    return snapshot

File: artifacts/api-server/src/routes/cameras.ts
import { Router, type IRouter } from "express";
import { db, camerasTable, analyticsTable, alertsTable } from "@workspace/db";
import { eq, desc, and, gte, sql } from "drizzle-orm";
import { generateDetectionSnapshot } from "../lib/simulation.js";

const router: IRouter = Router();

router.get("/cameras", async (_req, res) => {
  const cameras = await db.select().from(camerasTable).orderBy(camerasTable.id);
  res.json(cameras.map((c) => ({ ...c, ppeRequirements: c.ppeRequirements ?? [] })));
});

router.post("/cameras", async (req, res) => {
  const { name, location, streamUrl, ppeRequirements } = req.body as {
    name: string;
    location: string;
    streamUrl?: string;
    ppeRequirements: string[];
  };
  const [camera] = await db
    .insert(camerasTable)
    .values({ name, location, streamUrl, ppeRequirements: ppeRequirements ?? [] })
    .returning();
  res.status(201).json({ ...camera, ppeRequirements: camera.ppeRequirements ?? [] });
});

router.get("/cameras/:cameraId", async (req, res) => {
  const cameraId = parseInt(req.params.cameraId);
  const [camera] = await db.select().from(camerasTable).where(eq(camerasTable.id, cameraId));
  if (!camera) return res.status(404).json({ error: "Not found" });
  res.json({ ...camera, ppeRequirements: camera.ppeRequirements ?? [] });
});

router.patch("/cameras/:cameraId", async (req, res) => {
  const cameraId = parseInt(req.params.cameraId);
  const updates = req.body as Partial<{
    name: string;
    location: string;
    status: "active" | "inactive" | "error";
    streamUrl: string;
    ppeRequirements: string[];
  }>;
  const [camera] = await db
    .update(camerasTable)
    .set({ ...updates, updatedAt: new Date() })
    .where(eq(camerasTable.id, cameraId))
    .returning();
  if (!camera) return res.status(404).json({ error: "Not found" });
  res.json({ ...camera, ppeRequirements: camera.ppeRequirements ?? [] });
});

router.delete("/cameras/:cameraId", async (req, res) => {
  const cameraId = parseInt(req.params.cameraId);
  await db.delete(camerasTable).where(eq(camerasTable.id, cameraId));
  res.status(204).send();
});

router.get("/cameras/:cameraId/snapshot", async (req, res) => {
  const cameraId = parseInt(req.params.cameraId);
  const [camera] = await db.select().from(camerasTable).where(eq(camerasTable.id, cameraId));
  if (!camera) return res.status(404).json({ error: "Not found" });
  const snapshot = generateDetectionSnapshot(camera);
  res.json(snapshot);
});

export default router;
